fabric/src/main/java/me/lucko/luckperms/fabric/AbstractFabricBootstrap.java

```diff
@@ -0,0 +1,234 @@
+package me.lucko.luckperms.fabric;
+
+import com.mojang.authlib.GameProfile;
+import me.lucko.luckperms.common.dependencies.classloader.PluginClassLoader;
+import me.lucko.luckperms.common.plugin.bootstrap.LuckPermsBootstrap;
+import me.lucko.luckperms.common.plugin.logging.PluginLogger;
+import me.lucko.luckperms.common.plugin.scheduler.SchedulerAdapter;
+import net.fabricmc.fabric.api.event.server.ServerStartCallback;
+import net.fabricmc.fabric.api.event.server.ServerStopCallback;
+import net.fabricmc.loader.api.FabricLoader;
+import net.fabricmc.loader.launch.common.FabricLauncherBase;
+import net.luckperms.api.platform.Platform;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.network.ServerPlayerEntity;
+
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.stream.Stream;
+
+/**
+ * Since fabric can run on both a client and server, we have to leave some details of the environment to each platform's bootstrap.
+ */
+public abstract class AbstractFabricBootstrap implements LuckPermsBootstrap {
+
+    private static final String MODID = "luckperms";
+
+    public AbstractFabricBootstrap() {
+        this.classLoader = new FabricClassLoader();
+        this.schedulerAdapter = new FabricSchedulerAdapter(this);
+        this.plugin = new LPFabricPlugin(this);
+    }
+
+    /**
+     * The plugin logger
+     */
+    private PluginLogger logger = new FabricPluginLogger();
+
+    /**
+     * A scheduler adapter for the platform
+     */
+    private final SchedulerAdapter schedulerAdapter;
+
+    /**
+     * The plugin class loader.
+     */
+    private PluginClassLoader classLoader;
+
+    /**
+     * The plugin instance
+     */
+    private LPFabricPlugin plugin;
+
+    /**
+     * The time when the plugin was enabled
+     */
+    private Instant startTime;
+
+    // load/enable latches
+    private final CountDownLatch loadLatch = new CountDownLatch(1);
+    private final CountDownLatch enableLatch = new CountDownLatch(1);
+
+    @Override
+    public PluginLogger getPluginLogger() {
+        return this.logger;
+    }
+
+    @Override
+    public SchedulerAdapter getScheduler() {
+        return this.schedulerAdapter;
+    }
+
+    @Override
+    public PluginClassLoader getPluginClassLoader() {
+        return this.classLoader;
+    }
+
+    @Override
+    public CountDownLatch getLoadLatch() {
+        return this.loadLatch;
+    }
+
+    @Override
+    public CountDownLatch getEnableLatch() {
+        return this.enableLatch;
+    }
+
+    @Override
+    public String getVersion() {
+        // This should NEVER fail since this method is only invoked if LuckPerms is loaded.
+        return FabricLoader.getInstance().getModContainer(MODID).orElseThrow(RuntimeException::new).getMetadata().getVersion().getFriendlyString();
```

Comment:
Maybe the mod container should be stored as a field on this instance? Obtain it in the constructor?